File: app/src/main/java/edu/application/ui/adapters/LessonsRecyclerViewAdapter.java
package edu.application.ui.adapters;

import android.annotation.SuppressLint;
import android.app.Activity;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.navigation.Navigation;
import androidx.recyclerview.widget.RecyclerView;

import java.util.ArrayList;
import java.util.List;

import edu.application.R;
import edu.application.data.models.Lesson;
import edu.application.data.models.LessonQuestion;
import edu.application.databinding.FragmentLessonItemViewBinding;

public class LessonsRecyclerViewAdapter extends
        RecyclerView.Adapter<LessonsRecyclerViewAdapter.RecyclerViewItemViewHolder> {
    List<Lesson> data;

    public LessonsRecyclerViewAdapter() {
        this.data = new ArrayList<>();
    }


    public LessonsRecyclerViewAdapter(List<Lesson> data) {
        this.data = data;
    }

    @SuppressLint("NotifyDataSetChanged")
    public void updateData(List<Lesson> newData) {
        data = newData;

        notifyDataSetChanged();
    }

    @NonNull
    @Override
    public RecyclerViewItemViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        FragmentLessonItemViewBinding mBinding =
                FragmentLessonItemViewBinding.inflate(LayoutInflater.from(parent.getContext()));

        return new RecyclerViewItemViewHolder(mBinding.getRoot());
    }

    @SuppressLint("SetTextI18n")
    @Override
    public void onBindViewHolder(@NonNull RecyclerViewItemViewHolder holder, int position) {
        Lesson lesson = data.get(position);
        holder.binding.lessonTextName.setText(lesson.getTextName());
        holder.binding.lessonTextAuthor.setText(lesson.getTextAuthor());
        holder.binding.lessonAgeCategory.setText(
                String.valueOf(lesson.getAgeMin()) + '-' + lesson.getAgeMax());
        holder.binding.lessonTextSize.setText(lesson.getSize());

        holder.binding.getRoot().setOnClickListener(view -> {
            Bundle bundle = new Bundle();
            bundle.putString("text", lesson.getText());
            bundle.putInt("questions_count", lesson.getQuestions().size());
            for (int i = 0; i < lesson.getQuestions().size(); i++) {
                LessonQuestion question = lesson.getQuestions().get(i);
                bundle.putString("question" + i, question.getQuestion());
                bundle.putString("answer" + i, question.getAnswer());
            }
            Navigation.findNavController(
                    (Activity) holder.itemView.getContext(), R.id.nav_host_fragment)
                    .navigate(R.id.action_trainingFragment_to_lessonReadingFragment, bundle);
        });
    }

    @Override
    public int getItemCount() {
        return data.size();
    }

    static class RecyclerViewItemViewHolder extends RecyclerView.ViewHolder {
        public FragmentLessonItemViewBinding binding;

        public RecyclerViewItemViewHolder(@NonNull View itemView) {
            super(itemView);
            itemView.setLayoutParams(new RecyclerView.LayoutParams(
                    ViewGroup.LayoutParams.MATCH_PARENT,
                    ViewGroup.LayoutParams.WRAP_CONTENT));

            binding = FragmentLessonItemViewBinding.bind(itemView);
        }
    }
}

File: app/src/main/kotlin/com/application/ui/fragments/navigation/training/adapters/LessonsRecyclerViewAdapter.kt
package com.application.ui.fragments.navigation.training.adapters

import android.annotation.SuppressLint
import android.app.Activity
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.navigation.Navigation.findNavController
import androidx.recyclerview.widget.RecyclerView
import com.application.R
import com.application.data.models.Lesson
import com.application.data.models.LessonQuestion
import com.application.databinding.FragmentLessonItemViewBinding
import com.application.ui.fragments.navigation.training.adapters.LessonsRecyclerViewAdapter.RecyclerViewItemViewHolder

class LessonsRecyclerViewAdapter : RecyclerView.Adapter<RecyclerViewItemViewHolder>() {

    var data: List<Lesson> = ArrayList()

    @SuppressLint("NotifyDataSetChanged")
    fun updateData(newData: List<Lesson>) {
        data = newData

        notifyDataSetChanged()
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerViewItemViewHolder {
        val binding = FragmentLessonItemViewBinding.inflate(LayoutInflater.from(parent.context))

        return RecyclerViewItemViewHolder(binding.root)
    }

    @SuppressLint("SetTextI18n")
    override fun onBindViewHolder(holder: RecyclerViewItemViewHolder, position: Int) {
        val lesson = data[position]
        holder.binding.lessonTextName.text = lesson.textName
        holder.binding.lessonTextAuthor.text = lesson.textAuthor
        holder.binding.lessonAgeCategory.text = lesson.ageMin.toString() + '+'
        holder.binding.lessonTextSize.text = lesson.size

        holder.binding.root.setOnClickListener {
            val bundle = Bundle()
            bundle.putString("text", lesson.text)
            bundle.putInt("questions_count", lesson.questions.size)
            var question: LessonQuestion
            for (i in lesson.questions.indices) {
                question = lesson.questions[i]
                bundle.putString("question$i", question.question)
                bundle.putInt("answers_cnt$i", question.answers.size)
                for (j in question.answers.indices)
                    bundle.putString("answer${i}_$j", question.answers[j])
                bundle.putInt("correct_answer$i", question.correctAnswer)
            }
            findNavController((holder.itemView.context as Activity), R.id.nav_host_fragment)
                .navigate(R.id.action_trainingFragment_to_preparingFragment, bundle)
        }
    }

    override fun getItemCount(): Int = data.size

    class RecyclerViewItemViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {

        init {
            itemView.layoutParams = RecyclerView.LayoutParams(
                ViewGroup.LayoutParams.MATCH_PARENT,
                ViewGroup.LayoutParams.WRAP_CONTENT)

        }

        var binding: FragmentLessonItemViewBinding = FragmentLessonItemViewBinding.bind(itemView)
    }
}